File: sbt-plugin/src/main/scala/play/soap/sbtplugin/PlayClientGenerator.scala
/*
 * Copyright (C) 2015-2020 Lightbend Inc. <https://www.lightbend.com>
 */
package play.soap.sbtplugin

import java.io.Writer

import org.apache.cxf.tools.common.model.JavaPort
import org.apache.cxf.tools.util.ClassCollector
import org.apache.cxf.tools.wsdlto.frontend.jaxws.generators.ServiceGenerator

/**
 * Generator for the Play plugin
 */
class PlayClientGenerator extends ServiceGenerator with PlayGenerator {
  override def setCommonAttributes() = {
    super.setCommonAttributes()
    setPlayAttributes()
  }

  def setAttribute(name: String, value: AnyRef) = setAttributes(name, value)

  override def doWrite(templateName: String, outputs: Writer) = {
    // Override the template... it should only ever be sei.vm, but in case it's not.
    val newTemplate = if (templateName.endsWith("/service.vm")) {
      "play/soap/sbtplugin/client.vm"
    } else templateName

    setAttribute("portMethod", PortMethodNameGenerator)

    super.doWrite(newTemplate, outputs)
  }

  /**
   * Overridden to make the output name Scala instead of Java.
   */
  override def parseOutputName(packageName: String, filename: String) = {
    register(env.get(classOf[ClassCollector]), packageName, filename)
    parseOutputName(packageName, filename, ".scala")
  }

  private object PortMethodNameGenerator {
    def transform(port: JavaPort): String = {
      port.getName.head.toLower + port.getName.tail
    }
  }
}

File: plugin/src/main/java/play/soap/plugin/PlayClientGenerator.java
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */
package play.soap.plugin;

import java.io.Writer;
import org.apache.cxf.tools.common.ToolException;
import org.apache.cxf.tools.util.ClassCollector;
import org.apache.cxf.tools.wsdlto.frontend.jaxws.generators.ServiceGenerator;

/** Generator for the Play plugin */
public class PlayClientGenerator extends ServiceGenerator implements PlayGenerator {

  @Override
  protected void setCommonAttributes() {
    super.setCommonAttributes();
    setPlayAttributes();
  }

  @Override
  public void setAttribute(String name, Object value) {
    setAttributes(name, value);
  }

  @Override
  protected void doWrite(String templateName, Writer outputs) throws ToolException {
    // Override the template... it should only ever be sei.vm, but in case it's not.
    String newTemplate = templateName;
    if (templateName.endsWith("/service.vm")) {
      newTemplate = "play/soap/plugin/client.vm";
    }
    setAttribute("portMethod", new PlayPortMethodNameGenerator());
    super.doWrite(newTemplate, outputs);
  }

  /** Overridden to make the output name Scala instead of Java. */
  @Override
  protected Writer parseOutputName(String packageName, String filename) throws ToolException {
    register(env.get(ClassCollector.class), packageName, filename);
    return parseOutputName(packageName, filename, ".scala");
  }
}
